spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryExecutor.java

```diff
@@ -162,6 +171,27 @@ private SpannerStatementQueryExecutor() {
 		return sb.toString();
 	}
 
+	private static String fetchInterleaved(String sql,
+			SpannerPersistentEntity<?> persistentEntity, SpannerMappingContext mappingContext, boolean fetchInterleaved) {
+		if (!fetchInterleaved) {
+			return sql;
+		}
+
+		final String subquery = getChildrenSubquery(persistentEntity, mappingContext);
+		if (subquery.isEmpty()) {
+			return sql;
+		}
+
+		final Pair<String, String> parts = splitSelectAll(sql);
+		return parts.getSecond().isEmpty() ? sql : parts.getFirst() + subquery + parts.getSecond();
+	}
+
+	private static Pair<String, String> splitSelectAll(String sql) {
```

Comment:
Could you clarify what is this method for?

Reply:
The method name is probably bad, but...
The initial idea was to use wrapped query `SELECT * FROM (<initial query>)` in the same way as the method `applySort` at line 160. It was fine until we decided to use the `fetchInterleaved` as permanently `true`. After a number of failed integration tests and attempts to fix them, I understood that the wrapped query will not work. Mainly because of Spanner does not preserve sort ordering of embedded queries.
Then I concluded - when the user has created an SQL with a complete list of fields to select it is not really good idea to add some more fields with additional sub-selects - it could really affect performance to select fields that were not requested. But when the query is like `select * from Table ...` - it is the best candidate to split it into two parts an add interleaved fields into the middle. Like `select *, ` + <interlieved> + ` from Table ...` and it does not break any skip\limit or sorting conditions.

Reply:
I don't really like that we have to rely on `SELECT_ALL_PATTERN`, there could be cases where it doesn't work and that would lead to inconsistent behavior. It is not transparent and would be hard to detect. 

Is there a way to avoid regex matching? What if we use joins instead of sub-queries?

Reply:
To clarify, the query I'm thinking about would be something like that:
```
 SELECT     *
FROM       (SELECT * FROM   singers WHERE  singerid = 1) AS singers
INNER JOIN
           (
                    SELECT   singerid,
                             array_agg(struct(albumid, albumtitle, array
                             (
                                    select as struct singerid,
                                           albumid,
                                           albumtitle,
                                           array
                                           (
                                                  SELECT AS struct *
                                                  FROM   songs
                                                  WHERE  songs.singerid = albums.singerid
                                                  AND    songs.albumid = albums.albumid)))) AS alb
                    FROM     albums
                    GROUP BY albums.singerid) AS albums
ON         albums.singerid = singers.singerid
LIMIT      10
OFFSET     30 
```
where `SELECT * FROM   singers WHERE  singerid = 1` is the user-provided query.

What do you think about that approach?

Reply:
I understood
It looks more complex but the main - it is not clear how to process such resultset

Reply:
the new version is better :)

Reply:
here we have original SQL as a source for a new query - it is pointed in "... FROM (SELECT * FROM   singers WHERE  singerid = 1)". 
But the sort order of such SQL is not preserved

Reply:
Does it preserve order in current implementation? I believe we wrap the user-provided query in `SELECT * FROM (<query>) LIMIT .. OFFSET ..`

Reply:
No, it does not.
That is a major problem. :(
We have an integration test that will fail.
Some test from `SpannerRepositoryIntegrationTests#queryMethodsTest` line ~ 192 fails unpredictable on assertions `isGreaterThan` like
```
		List<Trade> buyTradesRetrieved = this.tradeRepository
				.annotatedTradesByAction("BUY");
		assertThat(buyTradesRetrieved).containsExactlyInAnyOrderElementsOf(trader1BuyTrades);
		assertThat(buyTradesRetrieved.get(0).getId()).isGreaterThan(buyTradesRetrieved.get(1).getId());
		assertThat(buyTradesRetrieved.get(1).getId()).isGreaterThan(buyTradesRetrieved.get(2).getId());
```

Reply:
We really wrap the user-provided queries by `SELECT * FROM (<query>) LIMIT .. OFFSET ..`. We doing it ONLY when we have a `SpannerPageableQueryOptions` parameter - in such case any "sorting" from it will overwrite the pointed by the SQL itself. And we should at least specify it in doc documentation I believe.

Reply:
I think we should just accept the fact that the sort order in custom `@Query`s will not be preserved. The solutions is to pass in the sort order in the `Pageable` argument. This is definitely something that needs to be in our refdoc.

Reply:
@s13o I think it's ok if ordering is not preserved. We apply ordering from `Pageable`, and that's how it should be passed. 

Could you roll back changes that introduced regexp?
I think after that we should be able to merge this PR.

Thanks

Reply:
oops, will try
but I'll have to change the test `SpannerRepositoryIntegrationTests#queryMethodsTest` line ~ 192

Reply:
May I ask you, why you don't like the idea to inject child-subquery after "select *"? It seems to be simple enough. I can't really imagine a case when it could fail. Right now - the only leading spaces can affect it. Ok, the regex could be like
```java
Pattern.compile("^\\s+?select(\\s+?all|\\s+?distinct)*(?<star>\\s+?\\*)", Pattern.CASE_INSENSITIVE);
```
The query is parsed only once, it is from an annotation, so we can cache the result and it should not affect performance.

Reply:
Sounds good. The sort order should be determined by the `Pageable` that is passed to the method. That is: 
```
PageRequest
				.of(2, 2, org.springframework.data.domain.Sort.by(Order.asc("id")))
```
lines 188-189

Reply:
but you didn't answer me about the regex. 
You probably don't like Pearl :)

Reply:
I need to have it merged so I'll revert the regex, but, please your opinion is very important to me.

Reply:
@s13o We avoid parsing the SQL in principle. Experience has taught us that it's a rather risky option. Even augmenting queries is risky, but we decided to go with it in the absence of better alternatives.
For example, note that Spanner SQL may contain [query hints](https://cloud.google.com/spanner/docs/query-syntax). That's one situation where parsing can fall apart.

Reply:
@s13o You don't have to merge it to undo a commit. I think `git revert <commit hash>` should do the trick.

Reply:
You goddamn right. I also hated the idea to parse the SQL from the beginning but saw no other alternative to satisfy failed IT tests.
Many thanks for the beautiful explanation.

Reply:
> @s13o You don't have to merge it to undo a commit. I think `git revert <commit hash>` should do the trick.

Finally, I found a time to complete the PR!
Sorry for the delay, please review.